src/solvers/refinement/string_constraint_generator.cpp

```diff
@@ -0,0 +1,2236 @@
+  const exprt &i, size_t max_size)
+{
+  string_exprt res(get_char_type());
+  typet type=i.type();
+  assert(type.id()==ID_signedbv);
+  exprt ten=from_integer(10, type);
+  exprt zero_char=constant_char('0');
+  exprt nine_char=constant_char('9');
+  exprt minus_char=constant_char('-');
+  exprt zero=refined_string_typet::index_zero();
+  exprt max=refined_string_typet::index_of_int(max_size);
+  // We add axioms:
+  // a1 : 0 <|res|<=max_size
+  // a2 : (res[0]='-')||('0'<=res[0]<='9')
+
+  and_exprt a1(res.is_strictly_longer_than(zero), res.axiom_for_is_shorter_than(max));
+  axioms.push_back(a1);
+
+  exprt chr=res[0];
+  exprt starts_with_minus=equal_exprt(chr, minus_char);
+  exprt starts_with_digit=and_exprt
+    (binary_relation_exprt(chr, ID_ge, zero_char),
+     binary_relation_exprt(chr, ID_le, nine_char));
+  or_exprt a2(starts_with_digit, starts_with_minus);
+  axioms.push_back(a2);
+
+  for(size_t size=1; size<=max_size; size++)
+  {
+    // For each possible size, we add axioms:
+    // all_numbers: forall 1<=i<=size. '0'<=res[i]<='9'
+    // a3 : |res|=size&&'0'<=res[0]<='9' =>
+    //      i=sum+str[0]-'0' && all_numbers
+    // a4 : |res|=size&&res[0]='-' => i=-sum
+    // a5 : size>1 => |res|=size&&'0'<=res[0]<='9' => res[0]!='0'
+    // a6 : size>1 => |res|=size&&res[0]'-' => res[1]!='0'
+    // a7 : size==max_size => i>1000000000
+    exprt sum=from_integer(0, type);
+    exprt all_numbers=true_exprt();
+    chr=res[0];
+    exprt first_value=typecast_exprt(minus_exprt(chr, zero_char), type);
+
+    for(size_t j=1; j<size; j++)
+    {
+      chr=res[j];
+      sum=plus_exprt(
+        mult_exprt(sum, ten),
+        typecast_exprt(minus_exprt(chr, zero_char), type));
+      first_value=mult_exprt(first_value, ten);
+      and_exprt is_number(
+        binary_relation_exprt(chr, ID_ge, zero_char),
+        binary_relation_exprt(chr, ID_le, nine_char));
+      all_numbers=and_exprt(all_numbers, is_number);
+    }
+
+    axioms.push_back(all_numbers);
+
+    equal_exprt premise=res.axiom_for_has_length(size);
+    equal_exprt constr3(i, plus_exprt(sum, first_value));
+    implies_exprt a3(and_exprt(premise, starts_with_digit), constr3);
+    axioms.push_back(a3);
+
+    implies_exprt a4(
+      and_exprt(premise, starts_with_minus),
+      equal_exprt(i, unary_minus_exprt(sum)));
+    axioms.push_back(a4);
+
+    // disallow 0s at the beginning
+    if(size>1)
+    {
+      equal_exprt r0_zero(res[zero], zero_char);
+      implies_exprt a5(
+        and_exprt(premise, starts_with_digit),
+        not_exprt(r0_zero));
+      axioms.push_back(a5);
+
+      exprt one=refined_string_typet::index_of_int(1);
+      equal_exprt r1_zero(res[one], zero_char);
+      implies_exprt a6(
+        and_exprt(premise, starts_with_minus),
+        not_exprt(r1_zero));
+      axioms.push_back(a6);
+    }
+
+    // we have to be careful when exceeding the maximal size of integers
+    // Warning this should be different depending on max size
```

Comment:
Actually this has already been dealt with so I removed the comment https://github.com/diffblue/cbmc/pull/374/commits/3223b168697f088979db450fe4678bd78530941c